mcs/class/System/System.Net.Sockets/Socket.cs

```diff
@@ -951,16 +951,33 @@ public bool AcceptAsync (SocketAsyncEventArgs e)
 					throw new InvalidOperationException ("AcceptSocket: The socket must not be bound or connected.");
 			}
 
-			e.curSocket = this;
-			e.Worker.Init (this, e, SocketOperation.Accept);
+			SocketAsyncResult sockares = new SocketAsyncResult (this, s => SocketAsyncEventArgsCallback (AcceptAsyncCallback, s), e, SocketOperation.Accept);
 
-			SocketAsyncResult sockares = e.Worker.result;
+			e.current_socket = this;
+			e.SetLastOperation (SocketOperationToSocketAsyncOperation (SocketOperation.Accept));
+			e.SocketError = SocketError.Success;
+			e.BytesTransferred = 0;
 
-			QueueSocketAsyncResult (readQ, e.Worker, sockares);
+			QueueIOSelectorJob (readQ, sockares.handle, new IOSelectorJob (IOOperation.Read, s => BeginAcceptCallback ((SocketAsyncResult) s), sockares));
```

Comment:
We can have a real no-allocation path whenever we switch the IO threadpool code to managed, as we have to pass this information in some way to native, and the simplest way is to have this `IOSelectorJob` object that is going to contain all the needed information.